src/rt/sched/threadsyncsystematic.h

```diff
@@ -0,0 +1,335 @@
+#include <mutex>
+
+/**
+ * This file contains the synchronisation implementation for suspending
+ * and resuming threads.  It is design to be a thin wrapper that could
+ * be replaced with alternatives for other platforms.
+ *
+ * This implementation provides systematic testing.
+ */
+namespace verona::rt
+{
+  enum class SystematicState
+  {
+    Active,
+    Finished
+  };
+
+  inline snmalloc::function_ref<bool()> true_thunk{[]() { return true; }};
+
+  template<typename T>
+  class ThreadSyncSystematic
+  {
+    /// Mutex for manipulating systematic testing datastructures
+    ///  * running_thread
+    ///  * systematic_status
+    ///  * shutdown
+    std::mutex m_sys;
+
+    /// Model underlying locking provided by the handle.
+    std::atomic<bool> m = false;
+
+    /// unpause incarnation
+    /// Complete wrap around will lead to lost wake-up.  This seems safe to
+    /// ignore.
+    std::atomic<size_t> unpause_incarnation = 0;
+
+    /// Specifies which thread is currently executing in systematic testing
+    /// nullptr is used to mean no thread is currently running.
+    T* running_thread = nullptr;
+
+    /// Specifies the first thread to consider when choosing a thread, if
+    /// being launch from outside.
+    T* first_thread = nullptr;
+
+    /// Used to prevent systematic testing attempt to access threads when the
+    /// runtime has been deallocated.
+    bool shutdown = true;
+
+    /// Must hold the systematic testing lock to call this.
+    /// External is used for calls from outside a scheduler thread.
+    void
+    choose_thread(std::unique_lock<std::mutex>&, T* me, bool external = false)
+    {
+      if (external && running_thread != nullptr)
+      {
+        // Runtime was not asleep, so don't try to wake up a thread.
+        return;
+      }
+
+      auto i = snmalloc::bits::ctz(Systematic::get_prng_next());
+      auto start = running_thread;
+      assert(me == start);
+      UNUSED(me);
+      if (start == nullptr)
+      {
+        // This is an external wake up
+        start = first_thread;
+      }
+
+      // Skip to a first choice for selecting.
+      while (i > 0)
+      {
+        start = start->next;
+        i--;
+      }
+
+      auto result = start;
+      while ((result->systematic_state != SystematicState::Active) ||
+             !result->guard())
+      {
+        result = result->next;
+        if (result == start)
+        {
+          // The following note is incase anyone tries to add an assertion about
+          // the conditions that could be true at this point.
+          //
+          // Note, this could have zero external event sources, with the runtime
+          // waking up from pausing.  The external thread that dropped the event
+          // source, can will still be `unpausing` the runtime.  So it appears
+          // all threads are going to sleep, but the external thread will
+          // continue to wake them. Until `runtime_pausing` is unset.

```

Comment:
```suggestion
          // The following note is for anyone wanting to add an assertion about
          // the conditions that could be true at this point.
          //
          // Note, this could have zero external event sources, with the runtime
          // waking up from pausing.  The external thread that dropped the event
          // source, can still be `unpausing` the runtime.  So it appears all
          // threads are going to sleep, but the external thread will continue to
          // wake them until `runtime_pausing` is unset.

```